Feature/dryad addon

kwierman opened the pull request:
Adding Dryad Addon. This time around, I've cleaned up the commits. I've got one squashed commit for my main changes and one for reverting all the small changes I used to get osf to work on my machine. Finally, I have a merge commit for automatically merging this with develop. This is now tested and passes flake8. Hopefully, this should work.

samchrisinger commented on website/addons/dryad/__init__.py:
```diff
@@ -0,0 +1,28 @@
+OWNERS = ['user', 'node']
+ADDED_DEFAULT = []
+ADDED_MANDATORY = []
+VIEWS = ['page']
```
Unless you plan to have a separate page for this addon (i.e. "Files", "Wiki", "Contributors" on the project dashboard) pages should be `[]`

samchrisinger commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,121 @@
+
+class AddonDryadUserSettings(AddonUserSettingsBase):
+
+    def delete(self, **kwargs):
```
Can this method be removed?

kwierman replied:
The documentation states that this method needs to be instantiated. I haven't tried deleting it, but if you think it's OK, then I will

samchrisinger replied:
If all the method does is call super, it can almost certainly be deleted.

samchrisinger commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,121 @@
+        A Dryad node is a collection of packages. Each package is specified by a DOI, and the title is saved automatically
+    """
+    dryad_package_doi = fields.StringField()
+    complete = True
```
There may be benefit to implementing this method. In general we use `complete` to assert that the addon has been authorized and the user has selected a container (I'm guessing `dryad_package_doi` here).

Likewise with `has_auth` (which you might call in `complete`)

samchrisinger replied:
^ I'm just now realizing `has_auth = True` may be totally accurate in the Dryad use case.

kwierman replied:
Yup. This is an example of methods that I've tried to remove in the past and which (the removal) causes the addon to break.

samchrisinger commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,121 @@
+from pyDryad import Dryad
+
+
+class AddonDryadUserSettings(AddonUserSettingsBase):
```
I saw this class definition and was at first very confused. 

Do you think you could add a docstring to this class briefly explaining the Dryad API's authentication model (or lack thereof)?

kwierman replied:
Sure thing. At the moment, there is no authentication. For this version of the addon, we're just pulling files from the public archive. For the next version, we'll implement authorization. I'm leaving this in because the documentation stated that **every** addon must implement User settings.

samchrisinger commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,121 @@
+        return ''
+
+    def serialize_waterbutler_credentials(self):
+        return {'storage': {}}
```
I'm thinking this should reference the linked Dryad package DOI

samchrisinger replied:
Never mind, this is done this below.

kwierman replied:
Yea, I wasn't sure what it put into the credentials and settings. As far as I can tell, these two are separated for convenience (or organization, whatever).  Since we're pulling in only the first version of this addon, I'm leaving them in.

samchrisinger commented on website/addons/dryad/routes.py:
```diff
@@ -0,0 +1,76 @@
+TEMPLATE_DIR = '../addons/dryad/templates/'
+
+
+page_routes = {'rules':
```
I'm seeing now that you do have code for a Dryad specific pages. At this I don't have a very clear idea of the expectations or featureset of this PR. Would you be willing to update the description with a more detailed description and screenshots of all UI changes?

kwierman replied:
Sure thing. Do you want this in the commit message? I can briefly write up what we've got right now in the PR messages.

samchrisinger replied:
The github PR description would be ideal. Thanks!

samchrisinger commented on website/addons/dryad/settings/defaults.py:
```diff
@@ -0,0 +1,32 @@
+
+#DRYAD OAI-PMH metadata harvesting settings
+#Note where requested, format as (date, prefix, dataset)
+DRYAD_OAI_IDENTIFY = "http://www.datadryad.org/oai/request?verb=Identify"
```
I saw you're requiring another python module to use as a Dryad API client, and am wondering if these constants are needed here?

kwierman replied:
Yup, you're right. The waterbutler folks had me remove the python Dryad module since the unit tests depend on using aiohttp very specifically. We can either remove these lines or do as WB does and remove the dependency.

samchrisinger commented on website/addons/dryad/static/node-cfg.js:
```diff
@@ -0,0 +1,44 @@
+
+                    }
+                    else{
+                        bootbox.alert("Failed to add DOI: "+doi+" to project. DOIs come in the form doi:10.5061/dryad.XXXX.");
```
Please log JS errors with Raven (e.g. https://github.com/CenterForOpenScience/osf.io/blob/04056a550b2585b01c01183f473f0c166c3b8a15/website/static/js/resources/base.js#L54)

samchrisinger commented on website/addons/dryad/static/node-cfg.js:
```diff
@@ -0,0 +1,44 @@
+	    		type: "GET",   
+    			url: check_url,   
+                data: {'doi' : doi},
+			    success : function(response) {
```
It might make sense to convert some of the views that don't need a rendered HTML response to use the `json_renderer`, `check_dryad_doi` being a good example of this. 

The only changes required would be changing the renderer in your routes.py and doing
`if (response.result === true)` here instead.

samchrisinger commented on website/addons/dryad/static/node-cfg.js:
```diff
@@ -0,0 +1,44 @@
+var AddonHelper = require('js/addonHelper');
+
+
+    $(document).ready(function() {
```
Please unindent this block and try to make sure indention level is consistent throughout this file.

samchrisinger commented on website/addons/dryad/static/node-cfg.js:
```diff
@@ -0,0 +1,44 @@
+
+    });
+
+$(window.contextVars.dryadSettingsSelector).on('submit', AddonHelper.onSubmitSettings);
```
Should this go in your `$(document).ready`?

samchrisinger commented on website/addons/dryad/views.py:
```diff
@@ -0,0 +1,225 @@
+from pyDryad import Dryad_DataOne, Dryad_SOLR
```
Not a blocker by any means, but the pep8 guidelines recommend no underscores in class names 

https://www.python.org/dev/peps/pep-0008/#class-names

samchrisinger commented on website/addons/dryad/views.py:
```diff
@@ -0,0 +1,225 @@
+    try:
+        m = d.metadata(doi)
+    except HTTPError:
+        return redirect("/project/{}".format(pid))
```
I'm a little confused by this redirect, but think a better understanding of the UI/feature set will help

samchrisinger commented on website/addons/dryad/views.py:
```diff
@@ -0,0 +1,225 @@
+
+    objectList = ET.Element("ul")
+
+    for obj in x.getElementsByTagName("objectInfo"):
```
It might be nice to move this code into a separate utility method for improved testability and readability.

samchrisinger commented on website/addons/dryad/views.py:
```diff
@@ -0,0 +1,225 @@
+            add_button.text = "Set Node Data to This"
+            add_button.attrib["href"] = "/project/{}/dryad/add?doi={}".format(pid, identifier)
+
+    ret.update({"content": ET.tostring(objectList)})
```
I would vasty prefer seeing HTML construction strictly limited to the template layer. Is there a reason this construction can't happen in a Mako template?

samchrisinger commented on framework/addons/data/addons.json:
```diff
@@ -150,6 +150,36 @@
+            },
+            "Delete files": {
+                "status": "NA",
+                "text": "Files deleted via OSF will not be deleted in Dryad."
```
Is this accurate?

kwierman replied:
Yup. The next release will feature pushing and pulling to Dryads staging and review databases, but for now, the only portions accessible will be the parts of the archive not under embargo.

samchrisinger replied:
But will deleting files be supported on the OSF interface at all?

kwierman replied:
In the next release it will. The current release does not.

lbanner commented:
Hi @kwierman I've got a request. It's challenging to review your work without a clear picture of your goals + expectations for this add-on and the OSF. Can you draft some documentation explaining such? But even more helpful, can you take screenshots from your local environment of what and how you have the add-on interacting and working for the user? This will be good to have for review, but also very helpful as we move into QA. Thanks!

kwierman commented:
Hey @lbanner @samchrisinger, I'm out of town at the moment, but I'm still resolving issues one at a time. I'll get right on to a guide as soon as I can.

lbanner commented:
TY @kwierman

kwierman commented:
# Dryad Addon Guide

## Intro to Dryad

[Dryad](http://datadryad.org/) is a data repository with a focus on biological and life sciences that provides professional curatorial services to researchers in order to meet the expectations of scientific journals. In many respects, Dryad can be treated identically to [Dataverse](http://dataverse.org/) and [Figshare](https://figshare.com/). Therefore, these two addons were used as the inspiration for the Dryad addon.

The primary difference between Dryad and the aforementioned other repositories is at the data curation stage. Data is submitted to Dryad via the web interface to a staging or _review_ data repository and arranged into a _package_. A package consists of a set of files and whatever metadata is available to the submitter. Once the submitter is satisfied with their package, the data is put under review and is available to Dryad curators. Curators may push the package back to the submitter for editing. The review-and-edit process may be repeated several times. Once the package passes curation, the package is moved to Dryad's _archive_.

The archive database and review space in Dryad are treated separately. As an external application, tracking a package from submission to archive is difficult due to no external identifier that can be used to identify the package in both submission and archive. Once in archive, a package __must__ have a _DOI_ issued by [Datacite](https://www.datacite.org/), and a package __may__ have this same DOI at the time of initial submission, but this is not guaranteed as multiple workflows allow for posteriori DOI creation.

The other challenge in implementing a Dryad addon is that only current API providing access to the submission process is the web API. You can find documentation on the full API [here](http://wiki.datadryad.org/Data_Access). This project was taken on knowing that Dryad's SWORD API needed completion. The SWORD API will be completed in January at the earliest and therefore some revisions have been made to the original proposal. Dataverse implements a similar API, however, some additional metadata will be provided through the API to account for the curatorial stages.

## Goals

The initial goal of this project was to provide access to Dryad's submission system through the OSF dashboard __and__ to provide access to Dryad's archive. Access to the submission system should allow users to prepare packages in other waterbutler-enabled providers and submit to Dryad's review space. Dryad's leadership has expressed a desire for this functionality because OSF provides the framework for users to organize their submissions a priori. Similarly, access to the archive provides a [good investment](http://datadryad.org/resource/doi:10.5061/dryad.j1fd7) for submitters.

### Changes to the goals

As the SWORD API work is currently in progress, the project was divided into two sections:

* v1 will provide access only to the archive (a read-only repository).
* v2 will incorporate the SWORD API changes.

This current PR is for the v1 addon. Some code additions from the changes for v2 have been left in the v1 PR either accidentally, or knowing (and occasionally suspecting) that life will be easier later with the code already in place. These changes will be noted in the documentation for the next  version development.

## Development Environment

The following tools were used in this development. Additionally, I have a recipe system in place for deploying all of this using a single-button interface.

* [OSF](https://github.com/kwierman/osf.io/tree/feature/dryad_addon)
    * Python 2.7.9
* [waterbutler](https://github.com/kwierman/waterbutler/tree/feature/dryad_provider)
    * Python 3.4.3
* [mfr](https://github.com/CenterForOpenScience/modular-file-renderer)
    * Python 3.4.3
* [pycas](https://github.com/kwierman/pycas)
    * Python 2.7.9

platform:
```
Linux 3.19.0-42-generic #48-Ubuntu SMP Thu Dec 17 22:54:45 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

Note that as of a couple weeks ago, I put the additional couple hours into fixing pycas so that it _should_ totally replicate fakecas



## General Workflow

As with the rest of the addons, Dryad can be enabled in the settings menu
![screenshot from 2016-01-07 10-21-47](https://cloud.githubusercontent.com/assets/1746808/12174208/609af3dc-b529-11e5-8cc3-8ec8424f0698.png)

Once enabled, the usual terms windows appears. As Dryad is (at the moment) read only, the terms should pretty much reflect this (at the moment). I just realized there's a typo in here. That'll be fixed.
![screenshot from 2016-01-07 10-29-18](https://cloud.githubusercontent.com/assets/1746808/12174288/b6a3a59e-b529-11e5-8a33-e9c4a75f931d.png)

Clicking __Confirm__ and the __Apply__ causes applies the addon to the current node and the Dryad Node settings panel appears in the settings page.

![screenshot from 2016-01-07 10-36-25](https://cloud.githubusercontent.com/assets/1746808/12174584/5cc4ff9e-b52b-11e5-9e92-31e363e3d2a9.png)

The Dryad addon uses the DOI field to access the archive. Initially this was chosen because one of the workflows users experience enables journals to send the DOI to users in this format. The other Dryad workflow requires Dryad to email users the DOI in this format. They can enter this information into the field here manually. This DOI is then verified against Dryad and if success, is saved to the node settings. The user is notified of this via fadeout message in the settings pane.

For users that haven't been emailed this information, a Dryad page has been crafted that allows browsing and searching the Dryad archive and setting the Node package to via simple button interface.

![screenshot from 2016-01-07 10-45-53](https://cloud.githubusercontent.com/assets/1746808/12174662/06497540-b52c-11e5-8a03-018060adc75e.png)

The above shows Dryad page loading. Please excuse the brutal overuse of the spinning loader icon, but it seems to flow nicely with the OSF aesthetic. This page was generated using webpack and is tied to a knockout ViewModel and hopefully applies to the coding style of the rest of OSF. Some minor changes will be made in the future, but for now Raven isn't reporting any problems. Dryad's API server goes down for about 5-10 minutes whenever a change is made to the base code, so I'm currently stress testing this to report back to the user when archive metadata cannot be obtained.

![screenshot from 2016-01-07 10-46-51](https://cloud.githubusercontent.com/assets/1746808/12174668/098ccc34-b52c-11e5-82b7-03e6e0a8a64d.png)

The above is the Dryad page loaded. The left hand panel is the available metadata of the current package. The archive is listed to the right. Additionally, users may search the archive using the search bar. While Dryad itself has a search function via the webportal, is does not have a browser function or the ability to one-button import the DOI to OSF.

![screenshot from 2016-01-07 10-46-59](https://cloud.githubusercontent.com/assets/1746808/12174671/0cded26a-b52c-11e5-8a5b-d7ac6f96696e.png)

Clicking on a package expands the panel to view the package metadata. 

![screenshot from 2016-01-07 10-52-58](https://cloud.githubusercontent.com/assets/1746808/12174857/115ff8b8-b52d-11e5-9a7d-9f9f78ea6ddc.png)

At the bottom of the metadata is a link to the Dryad details page and a button to set the Node data to this. The ViewModel automatically updates the Current Package Pane to reflect changes made here.

![screenshot from 2016-01-07 10-55-55](https://cloud.githubusercontent.com/assets/1746808/12174894/4691d3e4-b52d-11e5-8a70-336ff1c8396b.png)

Once the DOI has been set, the package and it's contents may be viewed in the file pane. Changes to the addon are noted in the log.

![screenshot from 2016-01-07 10-58-10](https://cloud.githubusercontent.com/assets/1746808/12174963/9933de76-b52d-11e5-9793-e36ee80c3e5a.png)

Files in Dryad can be viewed in mfr. The Fangorn settings are not finalized (for whatever odd reason that Delete button __cannot__ be removed. Additionally, some changes are being made to the watebutler provider so that the drag-and-drop interface works for copying files _out_ of Dryad but not _in_ to Dryad.

## Current Status

* v1
  * Some stress testing has not been completed. In the Python side, the code is very tolerant of 503 errors from Dryad. An additional field is being added to the JSON interface to allow the view models to inform the user when they need to wait for Dryad.
  * The typos mentioned above are being fixed.
  * Places in the code that are in place for the v2 upgrade are being documented with DO NOT REMOVE tags.
* v2
  * The SWORD API in the Dryad code base is being worked on as we speak... err... type.

The last point I'd like to address is the choice of using the full DOI in the format:

doi:10.5061/dryad.XXXX

This is the format that I received in the early discussions on Dryad. I'm not totally glued to it, but in the event that Dryad deploys an external tracking system that utilizes a similar format, I'd like to be as open as possible. For internal code's sake, we can sanitize this down in the knockout layer easily, but that decision should be made before releasing v1 in order to avoid unnecessary migrations.

Edit: Sorry, one last point. Would you be interesting in using Dryad to automatically site the data? We can use this addon also as a citation addon as the whole point of Dryad is curated data citation.

kwierman commented:
I'm going to go ahead and apoligize for the book length. The good news is that I think all of the requested changes were made. The bad news is that due to the nature of these changes, we're going to be looking at a bunch of totally new code. Git commits to come in couple of hours...

felliott commented:
@kwierman++!  I haven't read through this yet, but it'll be **very** helpful on the WaterButler side, too. Thank you!

Cheers,
Fitz

erinmayhood commented:
@kwierman I've been reviewing the UI and have a few comments. 

1. We'd like to keep the terms bootbox consistent for all add-ons. Can you please fill in the missing categories:

**Add/update files suggestion:** Dryad files are current read only and files cannot be added or updated.
**Delete files suggestion:** Dryad files cannot be deleted via OSF.
**Forking suggestion: **Forking a project or component does not copy Dryad authorization unless the user forking the project is the same user who authorized the Dryad add-on in the source project being forked.
**Registering suggestion: **Dryad content will be registered, but version history will not be copied to the registration.

2. In the Configure Add-ons section, please move content to the left to match other add-on formats. For example see:
![screen shot 2016-02-17 at 8 48 33 am](https://cloud.githubusercontent.com/assets/6423999/13111628/d5c5c0b0-d554-11e5-8052-32e37846c962.png)

3. And finally (and this might require discussion), we typically do not show add-ons in the primary navigation bar. Can the Dryad organization be understood this way:

- Packages are similar to folders
- Packages have metadata associated with them
- Objects live within packages
- Objects have metadata associated with them

Rather than having a separate Dryad page would something similar to the following work:

1. Could users successfully search for packages from within the files browser?
![screen shot 2016-02-17 at 9 22 00 am](https://cloud.githubusercontent.com/assets/6423999/13112236/f1aa1fa8-d557-11e5-9da4-a6141022cc08.png)

2. Could package metadata be viewable from the files browser (hover over, or another solution)?
 
3. When a user opens a package and selects an object, could the files pane be enhanced to accommodate the metadata?

Again, I realize this last point could involve more discussion.

kwierman commented:
@erinmayhood Sure thing. The folks over at Dryad were giving me some flak about including a page, so moving the page content to the file browser would probably be preferable. 

However, I'm not _entirely_ certain I can make searching Dryad from within the file browser possible, but I'm going to try and hack some Fangorn features together  to see what I can do without breaking the aesthetic. I'll get back to you on this in a bit.

One last thing I wanted to include was a citation widget for Dryad packages. This just forwards the citation from Dryad that users should include when publishing with Dryad content.

kwierman commented:
Hey @erinmayhood, So, I'm having trouble with a couple of things.
1. Viewing the package metadata by using hover over in the file browser is proving difficult.

2. Simple searches in the Dryad DB return something like a few thousand results. Putting this in the file browser slows the UI down considerably. I've tried to paginate the results, but from an addon perspective, the file browser wasn't really made for this.

In the meantime, I disabled the `'page'` feature of the addon. What I propose is that we move a bunch of that functionality to a more compact widget, and keep the page behind a link in the settings

![screenshot from 2016-02-23 16-33-37](https://cloud.githubusercontent.com/assets/1746808/13267445/4541f350-da4b-11e5-812a-8139c72a36e8.png)
So, we could have something like the above. I'll mock up a widget shortly so we can see if that works.

erinmayhood commented:
@kwierman I think a widget or a metadata pane would be great. I know @felliott is planning to follow up with you on this project and can help review the widget approach. Thanks!

felliott commented:
@kwierman, a heads-up: if you're planning on writing a dryad citation addon, @mfraezz has [refactored the citation addon code](https://github.com/CenterForOpenScience/osf.io/pull/4962). It hasn't been merged yet, but will probably go into an upcoming release.

kwierman commented:
@felliott Thanks!

kwierman commented:
Hey @erinmayhood ,

So, I moved all the UI elements over to a widget. The widget looks like:

![screenshot from 2016-03-02 13-45-41](https://cloud.githubusercontent.com/assets/1746808/13479151/11aeff48-e0a3-11e5-9cbb-9aaf41dc777a.png)

I split the widget into three panels. The first is the metadata panel.

![screenshot from 2016-03-02 13-45-49](https://cloud.githubusercontent.com/assets/1746808/13479181/29b7244e-e0a3-11e5-9294-8cb0305c6cbb.png)

The second is the browser panel:

![screenshot from 2016-03-02 13-45-57](https://cloud.githubusercontent.com/assets/1746808/13479191/341b9ed8-e0a3-11e5-9d4d-46c316124aaf.png)

The last is the citation panel:

![screenshot from 2016-03-02 13-46-04](https://cloud.githubusercontent.com/assets/1746808/13479195/3cbc62c0-e0a3-11e5-8d35-d5908d70e43f.png)

The citations here are scraped from Dryad. I'm told these are input to the archive as whole strings, so I'm passing this UI element straight from the Dryad servers. I'm not currently implementing the Dryad addon as a citation addon, but if you want, I can change that.

As usual, feedback is appreciated.

mfraezz commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,106 @@
+    dryad_package_doi = fields.StringField()
+    has_auth = True
+    provider_name = 'dryad'
+    connection = dryad_utils.DryadRepository()
```
It's somewhat of an anti-pattern to store initialized class-level objects. This really only matters when they store state, which this currently appears to not do. If it's expected that this might change with version 2 of this feature, consider instead being lazy:

```python
_api = None

@property
def api(self)
    if not self._api:
        self._api = dryad_utils.DryadRepository()
    return self._api
```

kwierman replied:
I like it. This might end up being replaced with an external provider, but for now, I wanted a proxy method for accessing the repository to make refactoring easier.

mfraezz commented on framework/addons/data/addons.json:
```diff
@@ -150,6 +150,37 @@
+                "text": "The OSF keeps track of changes you make to your Dryad packages through the OSF, but not for changes made using Dryad directly."
+            },
+            "Forking": {
+                "status": "none",
```
This says `none`, but the `text` reads like `partial`

mfraezz commented on framework/addons/data/addons.json:
```diff
@@ -150,6 +150,37 @@
+                "text": "Forking a project or component does not copy Dryad authorization unless the user forking the project is the same user who authorized the Dryad add-on in the source project being forked."
+            },
+            "Registering": {
+                "status": "none",
```
This says `none`, but the `text` reads like `partial`

mfraezz commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,106 @@
+    connection = dryad_utils.DryadRepository()
+
+    user_settings = fields.ForeignField(
+        'addondryadusersettings', backref='authorized'
```
We had a big push to remove backrefs several months ago, because they can lead to database locking issues. Could you remove this backref?

mfraezz replied:
If you use the more generalized base classes, you'll get this for free

mfraezz commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,106 @@
+        return {'storage': {}}
+
+    def serialize_waterbutler_settings(self):
+        ret = {}
```
Why not just `return {'doi': self.dryad_package_doi}` ?

mfraezz commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,106 @@
+    has_auth = True
+
+
+class AddonDryadNodeSettings(AddonNodeSettingsBase):
```
The switch to v2 using their (currently unstable) SWORD API will likely be simplified by using `AddonOAuthNodeSettingsBase`, maybe `StorageAddonBase` as well. This doesn't need to happen now, as it will require some extra finagling, but would simplify things down the line. Consider switching, but not a blocker

mfraezz commented on website/addons/dryad/settings/defaults.py:
```diff
@@ -0,0 +1,23 @@
+DRYAD_OAI_LISTMETADATAFORMAT = 'http://www.datadryad.org/oai/request?verb=ListMetadataFormats'
+DRYAD_OAI_LISTIDENTIFIERS = 'http://www.datadryad.org/oai/request?verb=ListIdentifiers&from={}&metadataPrefix={}&set={}'
+DRYAD_OAI_LISTRECORDS = 'http://www.datadryad.org/oai/request?verb=ListRecords&from={}&metadataPrefix={}&set={}'
+DRYAD_OAI_GETRECORD = 'www.datadryad.org/oai/request?verb=GetRecord&identifier={}&metadataPrefix={}'
```
Why is this the only one not prefixed with `r'https?://'` ?

mfraezz replied:
On a related note, why are some routes secure and others not?

mfraezz replied:
Also, consider building these urls from base urls, e.g.
```python
DRYAD_BASE_URL = 'http://www.datadryad.org/'

DRYAD_OAI_BASE = '{}oai/'.format(DRYAD_BASE_URL)
DRYAD_API_BASE = '{}mn/'.format(DRYAD_BASE_URL)
# etc.
```

Could also use `furl`, but that's not strictly necessary

mfraezz commented on website/addons/dryad/routes.py:
```diff
@@ -0,0 +1,85 @@
+            views.dryad_validate_doi,
+            json_renderer,
+        ),
+        Rule(
```
This rule and the one immediately below it should be a single route that accepts a put request, to set/clear the settings:

```python
Rule(
    [
        '/project/<pid>/dryad/settings/',
        '/project/<pid>/node/<nid>/dryad/settings/'
    ],
    'put',
    views.dryad_set_config,
    json_renderer,
),
```

mfraezz commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,106 @@
+        ret['urls'] = {}
+        ret.update({'dryad_package_doi': self.dryad_package_doi if self.dryad_package_doi else ''})
+        for endpoint in api_endpoints:
+            self.add_url_to_json(ret['urls'], endpoint)
```
This seems like something the serializer should be doing. See `website.addons.base.serializer.CitationsAddonSerializer`

mfraezz commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,106 @@
+    @property
+    def complete(self):
+        """
+            This should return self.dryad_package_doi is not None.
```
A `configured` property should be `return self.dryad_package_doi is not None`

mfraezz commented on website/addons/dryad/routes.py:
```diff
@@ -0,0 +1,85 @@
+from framework.routing import Rule, json_renderer
+from website.addons.dryad import views
+
+TEMPLATE_DIR = '../addons/dryad/templates/'
```
Is this used?

mfraezz commented on website/addons/dryad/static/dryadBrowser.js:
```diff
@@ -0,0 +1,199 @@
+    });
+}
+
+ViewModel.prototype.changeMessage = function(text, css, timeout) {
```
The `changeMessageMixin` is the preferred way of adding this functionality to viewmodels. See `s3UserConfig.js` or `accountSettings.js`

mfraezz commented on website/addons/dryad/static/dryadBrowser.js:
```diff
@@ -0,0 +1,199 @@
+          self.totalResults(response.total);
+          self.lastIndex(response.end);
+          ko.utils.arrayForEach(response.package_list,
+            function (d_pack) {
```
Prefer camelCase, as per the [devdocs](http://cosdev.readthedocs.io/en/latest/style_guides/javascript.html)

mfraezz commented on website/addons/dryad/static/dryadBrowser.js:
```diff
@@ -0,0 +1,199 @@
+    .done(function (response) {
+        self.firstIndex(response.start);
+        self.totalResults(response.total);
+        self.lastIndex(response.start+count);
```
Please add whitespace around mathematical operators.

mfraezz commented:
As discussed elsewhere, please try to get the node-settings UI working with `folderPicker`

mfraezz commented on website/addons/dryad/static/dryadNodeConfig.js:
```diff
@@ -0,0 +1,375 @@
+//moving everything over
+var waterbutler = require('js/waterbutler');
+
+function ViewModel(url) {
```
Consider `oop.extend`ing this from `OauthAddonFolderPicker`; it should give you some generic functionality out of the box, and also make it easier to provide a more consistent UX. See recent changes to `s3NodeConfig.js`.

mfraezz commented on website/addons/dryad/utils.py:
```diff
@@ -0,0 +1,272 @@
+        """
+        resp = self.connection.get(dryad_settings.DRYAD_WEB_API.format(doi))
+        # The below encoding is mandatory
+        return BeautifulSoup(resp.text.encode('utf-8'))
```
We tend to try to avoid `BeautifulSoup` because it's kind of… slow. Also, it means you’re parsing HTML, which means:
a) The file format could be changed any time as they redesign the front end, or,
b) You’re actually parsing an acrhival format that is xml, in which case `lxml` is preferred. It’s fast, purpose designed, and has tools for extracting data from files.

kwierman replied:
So, this was something of a problem that came up in the early stages of development. The citation information which is used in the licensing portion of this  addon has no API endpoint. In other words, the only way to grab this information is from the Web frontend. I'll see if I can try and reconstruct the citation from the metadata, but I coded this portion almost a year ago, so we'll see how this goes.

mfraezz commented on website/addons/dryad/utils.py:
```diff
@@ -0,0 +1,272 @@
+        using connection.
+    """
+
+    def get(self, url='localhost', params={}):
```
This shouldn't have a default `url`, it should probably just fail.

mfraezz commented on website/addons/dryad/utils.py:
```diff
@@ -0,0 +1,272 @@
+    def __init__(self, connection=DryadConnection()):
+        self.connection = connection
+
+    def get_dryad_metadata(self, doi='doi:10.5061/dryad.1850'):
```
None of these should have a default `doi`

mfraezz commented on website/addons/dryad/tests/test_models.py:
```diff
@@ -0,0 +1,45 @@
+from website.addons.dryad.utils import DryadRepository
+
+
+class DryadTestModel(DryadTestCase):
```
If you can get this working with the generic models, try to also use use the generic test mixins found in `website.addons.base.testing` -- they might require some overrides, but will give a large number of tests for free.

mfraezz commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,106 @@
+from website.addons.base import AddonUserSettingsBase, AddonNodeSettingsBase
+
+
+class AddonDryadUserSettings(AddonUserSettingsBase):
```
Minor, but preferred convention is `<addon_name>UserSettings`, rather than `Addon<addon_name>UserSettings`. Same goes for node

mfraezz commented:
Somewhat light first pass complete. Going into much more depth at this stage would have diminishing returns. I'll do a more thorough pass after most of the suggested changes are in. ❄️ 

Feel free to reach out to me, either here or on Slack, for questions or clarification. Good work so far; I can tell the frontend/UI took a lot of effort, and after reading reading up on the Dryad API... I'm sorry. Back to you, @kwierman 🔙

kwierman commented:
Assuming the next batch of commits pass the tests :pray: , I think  I'm ready to pass this back to you guys. The major improvements here is implementing the citation base class in the python layer and the folderpicker base class in the js layer. Let me know what you think. Now that my dissertation and defense are both out of the way, the turnaround on subsequent commits should be MUCH shorter.

Anyway, let me know what you think. I think I hit up all of your comments, but if I missed something, I'll get right on it. :back:

mfraezz commented on website/addons/dryad/templates/dryad_user_settings.mako:
```diff
@@ -0,0 +1,11 @@
+<!-- Authorization -->
```
This file doesn't need to exist until the Dryad API is updated.

mfraezz commented on website/addons/dryad/templates/log_templates.mako:
```diff
@@ -0,0 +1,7 @@
+<script type="text/html" id="dryad_doi_set">
```
This file doesn't need to exist -- it should now be handled by the `.json` logs.

mfraezz commented on website/addons/dryad/tests/test_utils.py:
```diff
@@ -0,0 +1,35 @@
+        assert_true(len(resp.getElementsByTagName("identifier")) == 20)
+
+    def test_get_dryad_search_results_json_formatted(self):
+        # TODO. This has yet to be implemented
```
Is this still true?

felliott commented on website/addons/dryad/static/dryadNodeConfig.js:
```diff
@@ -0,0 +1,508 @@
+'use strict';
+
+var ko = require('knockout');
+require('knockout.punches');
```
The interface isn't working for me because `knockout.punches` is missing.  We took that out in v0.72.

felliott commented on website/addons/dryad/provider.py:
```diff
@@ -0,0 +1,337 @@
+        :returns:  bool -- True if the doi is found in the dryad archive.
+        :raises: urllib.error.HTTPError
+        """
+        dryad_metadata_path = furl(DRYAD_BASE_URL)
```
This check is failing for me when doing my local testing.  I think the base url for this check (and the one in `get_dryad_metadata`) changed to `http://api.datadryad.org/`

felliott commented on website/addons/dryad/model.py:
```diff
@@ -0,0 +1,84 @@
+    provider_name = 'dryad'
+    oauth_provider = DryadProvider
+    serializer = DryadSerializer
+    dryad_package_doi = fields.StringField()
```
@kwierman: the `validate_doi` validator doesn't like the `doi:` prefix you've been saving.  You'll need to strip that off and update the provider to prepend that before issuing your requests.

felliott commented on website/addons/dryad/static/dryadLogActionList.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "dryad_doi_set" : "${user} linked the Dryad package ${doi} to ${node}",
```
The variables available to these templates are set in `website/static/js/logTextParser.js`.  `${user}` and `${node}` should already be defined, but you will probably need to create a parser for `${doi}`.  For an example, see the [`${repo}` variable](https://github.com/CenterForOpenScience/osf.io/blob/develop/website/addons/github/static/githubLogActionList.json#L2) in the GitHub log actions.  The code to support that is [here](https://github.com/CenterForOpenScience/osf.io/blob/develop/website/static/js/logTextParser.js#L487) and relies on variables exposed through the Node Logs API [here](https://github.com/CenterForOpenScience/osf.io/blob/develop-backup/api/logs/serializers.py#L56).

felliott commented on website/addons/dryad/static/dryadNodeConfig.js:
```diff
@@ -0,0 +1,508 @@
+        var self = this;
+        var xhr = new XMLHttpRequest();
+        xhr.open("PUT", waterbutler.buildUploadUrl(path, 'osfstorage', self.node_id, {
+            name: 'DRYAD_LICENSE.md'
```
This might also need a `kind: 'file',`

mfraezz commented:
Closing in favor of #7417 , consider this merged. Thanks for all your hard work, @kwierman :octocat:
